### PR TITLE
DOC: MLHandler documentation

### PR DESCRIPTION
**Pull request by jaidevd**
Describe primary operations currently supported by MLHandler. Not to be merged before gramener/gramex#336

### CODE REVIEW

File: mlhandler/README.md [hunk 1]
```diff
@@ -0,0 +1,505 @@
...
+  kwargs:
+    data: $YAMLPATH/titanic.csv  # Path to the training dataset
+    model:
+      # The classification or regression algorithm to use
+      class: LogisticRegression
+
+      # Location where the trained model will be saved
+      path: $YAMLPATH/titanic.pkl
+
+      # The target column
```
Comment (jaidevd) [inline mlhandler/README.md:206]: Comment isn't obvious.

File: mlhandler/README.md [hunk 2]
```diff
@@ -0,0 +1,505 @@
...
+4. one-hot encode the categorical columns and normalize any remaining columns -
+   which are implicitly assumed to be numerical
+5. train the model and save it.
+
+Other kwargs that are supported are:
+
+* `include`: List of columns to include for training. If `include` and `exclude`
+           are both specified, the latter is ignored.
+* `nums`: List of numerical features which will be normalized.
+* `dopna`: Whether to drop NAs from the training data - true by default. If set
```
Comment (jaidevd) [inline mlhandler/README.md:230]: Typo.

File: mlhandler/README.md [hunk 3]
```diff
@@ -0,0 +1,505 @@
...
+2. load the training data as a pandas dataframe
+3. drop the excluded columns
+4. one-hot encode the categorical columns and normalize any remaining columns -
+   which are implicitly assumed to be numerical
+5. train the model and save it.
+
+Other kwargs that are supported are:
+
+* `include`: List of columns to include for training. If `include` and `exclude`
+           are both specified, the latter is ignored.
```
Comment (jaidevd) [inline mlhandler/README.md:228]: "If both are specified, include overrides exclude."

File: mlhandler/README.md [hunk 4]
```diff
@@ -0,0 +1,505 @@
...
+* `nums`: List of numerical features which will be normalized.
```
Comment (jaidevd) [inline mlhandler/README.md:229]: "List of nums. All nums are normalized with StandardScaler"

File: mlhandler/README.md [hunk 5]
```diff
@@ -0,0 +1,505 @@
...
+    </code></pre>
+  </div>
+  <div class="tab-pane fade" id="curlcode-bulk" role="tabpanel" aria-labelledby="curl-bulk-tab">
+    <pre><code>
+    curl -X POST -d @titanic.xlsx 'mlhandler?_action=predict'
+    </code></pre>
+  </div>
+</div>
+
+# Creating New Models
```
Comment (jaidevd) [inline mlhandler/README.md:189]: This comes before exposing existing models.

File: mlhandler/README.md [hunk 6]
```diff
@@ -0,0 +1,505 @@
...
+    <pre><code>
+    >>> import requests
+    >>> requests.get('mlhandler?Sex=male&Age=22&SibSp=1&Parch=0&Fare=7.25&Pclass=3&Embarked=S')
+    </code></pre>
+  </div>
+  <div class="tab-pane fade" id="ajaxcode" role="tabpanel" aria-labelledby="ajax-tab">
+    <pre><code>
+    $.ajax({
+    	url: 'mlhandler?Sex=male&Age=22&SibSp=1&Parch=0&Fare=7.25&Pclass=3&Embarked=S',
+      method: 'GET'
```
Comment (jaidevd) [inline mlhandler/README.md:120]: Indentation

File: mlhandler/README.md [hunk 7]
```diff
@@ -0,0 +1,505 @@
...
+  <div class="tab-pane fade" id="curlcode" role="tabpanel" aria-labelledby="curl-tab">
+    <pre><code>
+    curl -X GET mlhandler?Sex=male&Age=22&SibSp=1&Parch=0&Fare=7.25&Pclass=3&Embarked=S'
+    </code></pre>
+  </div>
+</div>
+
+
+Note that the URL parameters in the GET query are expected to be fields in the
+training dataset.
```
Comment (jaidevd) [inline mlhandler/README.md:133]: Replace with: The params can be passed through a Python dict or a JS object.

File: mlhandler/README.md [hunk 8]
```diff
@@ -0,0 +1,505 @@
...
+  <div class="tab-pane fade show active" id="pycode-bulk" role="tabpanel" aria-labelledby="python-bulk-tab">
+    <pre><code>
+    >>> import requests
+    >>> requests.post('mlhandler?_action=predict', files={'file': open('titanic.xlsx', 'rb')})
+    </code></pre>
+  </div>
+  <div class="tab-pane fade" id="ajaxcode-bulk" role="tabpanel" aria-labelledby="ajax-bulk-tab">
+    <pre><code>
+    $.ajax({
+    	url: 'mlhandler?_action=predict',
```
Comment (jaidevd) [inline mlhandler/README.md:174]: Check indentation.

File: mlhandler/README.md [hunk 9]
```diff
@@ -0,0 +1,505 @@
...
+
+
+# Model operations
+
+## Retraining the model
+An existing model can be retrained by POSTing data and specifying a target
+column. To do so, we need to:
+
+1. post the training data as JSON records
+2. set the `_action` query parameter to `retrain`
```
Comment (jaidevd) [inline mlhandler/README.md:253]: rephrase as "set `?_action=retrain`", also for the line below

Comment (jaidevd) [reply mlhandler/README.md:253]: Mention that query params can be dicts, as above.

File: mlhandler/README.md [hunk 10]
```diff
@@ -0,0 +1,505 @@
...
+modified with a PUT request. The following request replaces the logistic
+regression earlier with a random forest classifier:
+
+```bash
+PUT /mlhandler?_model&class=RandomForestClassififer
+```
+
+Note that at this stage, the model has simply been replaced, but _not_
+retrained. To train it, we can POST to it with `?_action=retrain` parameter as
+mentoined above.
```
Comment (jaidevd) [inline mlhandler/README.md:408]: Typo. Also, add an example.

File: mlhandler/README.md [hunk 11]
```diff
@@ -0,0 +1,505 @@
...
+
+Similarly, any parameter of the model can be changed. For example, to change the
+default number of estimators used in a random forest classifier (100), use:
```
Comment (jaidevd) [inline mlhandler/README.md:411]: Not clear that 100 is the default.

File: mlhandler/README.md [hunk 12]
```diff
@@ -0,0 +1,505 @@
...
+In general, the model's class and any of its parameters can be chained together
+in the PUT request. For example, to change the model to an SGDClassifier with a
+log loss, use:
+
+```bash
+PUT /mlhandler?_model&class=SGDClassifier&loss=log
+```
+
+Any query parameter except `class` which matches the signature of the
+[SGDClassifier](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.SGDClassifier.html) will be coerced into the model.
```
Comment (jaidevd) [inline mlhandler/README.md:426]: Not clear.

File: mlhandler/README.md [hunk 13]
```diff
@@ -0,0 +1,505 @@
...
+To remove the serialized model from the disk and disable further operations, use
+a delete request as follows:
+
+```bash
+DELETE /mlhandler?_model
+```
+
+# MLHandler Templates
+
+MLHandler supports Tornado templates. You can specify a `template` kwarg as follows:
```
Comment (jaidevd) [inline mlhandler/README.md:447]: Why would you use templates?

File: mlhandler/README.md [hunk 14]
```diff
@@ -0,0 +1,505 @@
...
+  handler: MLHandler
+  template: $YAMLPATH/template.html
+```
+
+The template so specified will be rendered at the `pattern` URL, and it will
+have access to the following variables:
+
+* `{{ handler }}`: The MLHandler instance
+* `{{ handler.model }}`: The sklearn object / estimator / model
+* `{{ data }}`: The training dataset, available as a pandas DataFrame.
```
Comment (jaidevd) [inline mlhandler/README.md:461]: Mention .__class__.__name__ and other estimator attributes included in the template. Mention that these are sklearn objects, but call out the important attributes of those.
